[Freshness] US-1953530: Refresh top articles by PV - .NET and Languages Content (3) (#30093)
dotnet/docs: docs/core/testing/index.md

description: This article gives a brief overview of testing concepts, terminology, and tools for testing in .NET.
author: IEvangelist
ms.author: dapine
ms.date: 07/18/2022
ms.custom: devdivchpfy22
---

# Testing in .NET

> This article introduces the concept of testing and illustrates how different kinds of tests can be used to validate code. There are various tools available for testing .NET applications, such as the [.NET CLI](#net-cli) or [Integrated Development Environments (IDEs)](#ide).

Author: Ok, updated the content.

## Test types

> Having automated tests is a great way to ensure that the application code does what its authors intend it to do. This article covers unit tests, integration tests, and load tests.

Author: Ok, updated the content.

### Unit tests

> A *unit test* is a test that exercises individual software components or methods, also known as "unit of work." Unit tests should only test code within the developer's control. They don't test infrastructure concerns. Infrastructure concerns include interacting with databases, file systems, and network resources.

Author: Ok, updated it.

### Load tests

A *load test* aims to determine whether or not a system can handle a specified load. For example, the number of concurrent users using an application and the app's ability to handle interactions responsively. For more information on load testing of web applications, see [ASP.NET Core load/stress testing](/aspnet/core/test/load-tests).

## Test considerations

> Keep in mind there are [best practices](unit-testing-best-practices.md) for writing tests. For example, [Test Driven Development (TDD)](https://deviq.com/test-driven-development) is when a unit test is written before the code it's meant to check. TDD is like creating an outline for a book before you write it. It's meant to help developers write simpler, readable, and efficient code.

Author: Ok, updated it.

## Testing tools

.NET is a multi-language development platform, and you can write various test types for [C#](../../csharp/index.yml), [F#](../../fsharp/index.yml), and [Visual Basic](../../visual-basic/index.yml). For each of these languages, you can choose between several test frameworks.

### xUnit

> [xUnit](https://xunit.net) is a free, open-source, community-focused unit testing tool for .NET. The original inventor of NUnit v2 writes xUnit.net. xUnit.net is the latest technology for unit testing .NET apps. It also works with ReSharper, CodeRush, TestDriven.NET, and [Xamarin](https://dotnet.microsoft.com/apps/xamarin). It's a project of the [.NET Foundation](https://dotnetfoundation.org) and operates under it's code of conduct.

Reviewer: ```suggestion
[xUnit](https://xunit.net) is a free, open-source, community-focused unit testing tool for .NET. The original inventor of NUnit v2 wrote xUnit.net. xUnit.net is the latest technology for unit testing .NET apps. It works with ReSharper, CodeRush, TestDriven.NET, and [Xamarin](https://dotnet.microsoft.com/apps/xamarin). xUnit.net is a project of the [.NET Foundation](https://dotnetfoundation.org) and operates under its code of conduct.
```

Author: Ok, updated the content.

> [xUnit](https://xunit.net) is a free, open source, community-focused unit testing tool for .NET. The original inventor of NUnit v2 writes xUnit.net. xUnit.net is the latest technology for unit testing .NET apps. It also works with ReSharper, CodeRush, TestDriven.NET, and [Xamarin](https://dotnet.microsoft.com/apps/xamarin). It's a project of the [.NET Foundation](https://dotnetfoundation.org) and operates under their code of conduct.

Author: Ok, updated it.

### NUnit

> [NUnit](https://nunit.org) is a unit-testing framework for all .NET languages. Initially NUnit is ported from JUnit, the current production release has been rewritten with many new features and support for a wide range of .NET platforms. It's a project of the [.NET Foundation](https://dotnetfoundation.org).

Author: Ok, updated it.

### .NET CLI

> You can run a solutions unit test from the [.NET CLI](../tools/index.md), with the [dotnet test](../tools/dotnet-test.md) command. The .NET CLI exposes most of the functionality that [Integrated Development Environments (IDEs)](#ide) make available through user interfaces. The .NET CLI is cross-platform and available to use as part of continuous integration and delivery pipelines. The .NET CLI is used with scripted processes to automate common tasks.

Author: Ok, updated it.
